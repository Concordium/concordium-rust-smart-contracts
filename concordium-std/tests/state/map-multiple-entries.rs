//! This test checks that you cannot have multiple entries from a `StateMap`
//! alive at the same time.
//!
//! When compiling it, the borrow-checker is supposed to throw an error.
use concordium_std::*;

pub fn main() {
    let mut state_builder = StateBuilder::open(ExternStateApi::open());
    let mut map: StateMap<u8, u8, _> = state_builder.new_map();
    // Get two entries.
    let e1 = map.entry(0u8);
    let e2 = map.entry(1u8);
    // Use them, so we are certain that their lifetimes overlap.
    e1.or_insert(1);
    e2.or_insert(2);
}
