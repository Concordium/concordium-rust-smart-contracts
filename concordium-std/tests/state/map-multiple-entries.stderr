error[E0499]: cannot borrow `map` as mutable more than once at a time
  --> tests/state/map-multiple-entries.rs:12:14
   |
11 |     let e1 = map.entry(0u8);
   |              --- first mutable borrow occurs here
12 |     let e2 = map.entry(1u8);
   |              ^^^ second mutable borrow occurs here
13 |     // Use them, so we are certain that their lifetimes overlap.
14 |     e1.or_insert(1);
   |     -- first borrow later used here
